### ISSUE TITLE
PROJECT4

### ISSUE THREAD

**Issue by KARL961709** [KARL961709/CodeKarl96#8]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.linear_model import LinearRegression
import warnings
warnings.filterwarnings('ignore')

# ==================== DATOS ====================
datos = {
    'fecha': ['feb24', 'mar24', 'abr24', 'may24', 'jun24', 'jul24', 'ago24', 
              'sep24', 'oct24', 'nov25', 'dic24', 'ene25', 'feb25', 'mar25', 
              'abr25', 'may25'],
    'PTO_INV': [15616, 15765, 15895, 16050, 16277, 16268, 16444, 16638, 
                16670, 16823, 16866, 16891, 16939, 17037, 17132, 17262],
    'FACTURACION': [297.9, 361.2, 380.1, 383.3, 371.4, 336.4, 380.5, 
                    415, 462.6, 382.5, 417.3, 426.5, 460, 530.5, 449, 392.5],
    'EAD': [15597, 15806, 15920, 16096, 16317, 16328, 16540, 16670, 
            16951, 17177, 17262, 17442, 17695, 17962, 18041, 17986]
}

df = pd.DataFrame(datos)

print("="*80)
print("ANÁLISIS ALTERNATIVO: CAPTURANDO CAMBIOS DE TENDENCIA")
print("="*80)

# ==================== 1. ANÁLISIS DE CAMBIO DE TENDENCIA ====================
print("\n1. ANÁLISIS DEL COMPORTAMIENTO RECIENTE")
print("-"*60)

# Calcular cambios mes a mes
df['cambio_EAD'] = df['EAD'].diff()
df['cambio_porcentual'] = df['EAD'].pct_change() * 100

print("\nÚltimos 6 períodos:")
print(df[['fecha', 'EAD', 'cambio_EAD', 'cambio_porcentual']].tail(6))

# Identificar punto de inflexión
print("\n2. IDENTIFICACIÓN DE PUNTO DE INFLEXIÓN")
print("-"*60)

# Análisis de ventanas móviles
window_size = 3
df['media_movil_3'] = df['EAD'].rolling(window=window_size).mean()
df['tendencia_local'] = df['media_movil_3'].diff()

print("\nTendencia local (cambio en media móvil):")
print(df[['fecha', 'EAD', 'media_movil_3', 'tendencia_local']].tail(8))

# ==================== 2. ANÁLISIS POR SEGMENTOS ====================
print("\n3. ANÁLISIS POR SEGMENTOS TEMPORALES")
print("-"*60)

# Dividir en segmentos
segmento1 = df[:10]  # feb24 - nov24
segmento2 = df[10:]  # dic24 - may25

# Tendencia Segmento 1
x1 = np.arange(len(segmento1))
slope1, intercept1 = np.polyfit(x1, segmento1['EAD'].values, 1)
print(f"\nSegmento 1 (feb24-nov24):")
print(f"  Pendiente: {slope1:.2f} unidades/mes")
print(f"  Crecimiento mensual promedio: {segmento1['EAD'].pct_change().mean()*100:.2f}%")

# Tendencia Segmento 2
x2 = np.arange(len(segmento2))
slope2, intercept2 = np.polyfit(x2, segmento2['EAD'].values, 1)
print(f"\nSegmento 2 (dic24-may25):")
print(f"  Pendiente: {slope2:.2f} unidades/mes")
print(f"  Crecimiento mensual promedio: {segmento2['EAD'].pct_change().mean()*100:.2f}%")

# ==================== 3. MODELOS ALTERNATIVOS ====================
print("\n4. MODELOS ALTERNATIVOS CONSIDERANDO CAMBIO DE TENDENCIA")
print("-"*60)

# Preparar datos para predicción
future_months = ['jun25', 'jul25', 'ago25', 'sep25', 'oct25', 'nov25', 'dic25']
predictions_dict = {}

# Método 1: Tendencia de los últimos N meses
def predict_recent_trend(df, n_months, n_future):
    recent_data = df.tail(n_months)
    x = np.arange(len(recent_data))
    slope, intercept = np.polyfit(x, recent_data['EAD'].values, 1)
    
    predictions = []
    for i in range(n_future):
        pred = slope * (len(recent_data) + i) + intercept
        predictions.append(pred)
    return predictions, slope

# Probar con diferentes ventanas
for n in [3, 4, 6]:
    preds, slope = predict_recent_trend(df, n, 7)
    predictions_dict[f'Tendencia_{n}M'] = preds
    print(f"\nTendencia últimos {n} meses:")
    print(f"  Pendiente: {slope:.2f}")
    print(f"  Predicción jun25: {preds[0]:.0f}")
    print(f"  Predicción dic25: {preds[-1]:.0f}")

# Método 2: Modelo de cambio de régimen
print("\n\nMODELO DE CAMBIO DE RÉGIMEN:")
# Usar tendencia reciente con ajuste por desaceleración
recent_changes = df['cambio_EAD'].tail(4).values
avg_recent_change = np.mean(recent_changes)
decay_factor = 0.9  # Factor de desaceleración

predictions_regime = []
last_value = df['EAD'].iloc[-1]
current_change = avg_recent_change

for i in range(7):
    # Aplicar desaceleración gradual
    current_change = current_change * decay_factor
    pred = last_value + current_change
    predictions_regime.append(pred)
    last_value = pred

predictions_dict['Cambio_Regimen'] = predictions_regime

# Método 3: Modelo cíclico
print("\n\nMODELO CÍCLICO:")
# Detectar patrón cíclico
cycle_length = 12  # Asumiendo ciclo anual
position_in_cycle = len(df) % cycle_length

predictions_cyclic = []
last_value = df['EAD'].iloc[-1]

for i in range(7):
    # Factor estacional basado en posición en el ciclo
    month_position = (position_in_cycle + i) % 12
    seasonal_factor = 1 + 0.02 * np.sin(2 * np.pi * month_position / 12)
    
    # Tendencia base moderada
    base_growth = 0.002  # 0.2% mensual
    pred = last_value * (1 + base_growth) * seasonal_factor
    predictions_cyclic.append(pred)
    last_value = pred

predictions_dict['Modelo_Ciclico'] = predictions_cyclic

# Método 4: Escenario conservador (estabilización)
print("\n\nESCENARIO CONSERVADOR (ESTABILIZACIÓN):")
# Asumir que EAD se estabiliza cerca del último valor
base_value = df['EAD'].iloc[-1]
small_growth = 0.001  # 0.1% mensual

predictions_conservative = []
for i in range(7):
    pred = base_value * (1 + small_growth) ** (i + 1)
    predictions_conservative.append(pred)

predictions_dict['Conservador'] = predictions_conservative

# ==================== 4. VISUALIZACIÓN COMPARATIVA ====================
plt.figure(figsize=(16, 10))

# Datos históricos
n_hist = len(df)
x_hist = np.arange(n_hist)
x_future = np.arange(n_hist, n_hist + 7)

# Plot histórico
plt.plot(x_hist, df['EAD'], 'o-', color='navy', linewidth=2.5, 
         markersize=7, label='Histórico', zorder=5)

# Marcar el cambio de tendencia
plt.axvline(x=10.5, color='red', linestyle='--', alpha=0.5, 
            label='Cambio de tendencia')

# Área de proyección
plt.axvspan(n_hist, n_hist + 6, alpha=0.1, color='yellow')

# Plotear diferentes escenarios
colors = ['green', 'orange', 'red', 'purple', 'brown', 'pink']
linestyles = ['-', '--', ':', '-.', '-', '--']

for i, (method, preds) in enumerate(predictions_dict.items()):
    plt.plot(x_future, preds, linestyles[i % len(linestyles)], 
             color=colors[i % len(colors)], linewidth=2, 
             label=method, alpha=0.8)

# Agregar anotaciones para los últimos puntos históricos
for i in range(-3, 0):
    plt.annotate(f'{df["EAD"].iloc[i]:.0f}', 
                xy=(n_hist + i, df['EAD'].iloc[i]), 
                xytext=(0, 10), textcoords='offset points',
                ha='center', fontsize=8)

# Configuración
all_periods = list(df['fecha']) + future_months
plt.xticks(range(0, len(all_periods), 2), 
           [all_periods[i] for i in range(0, len(all_periods), 2)], 
           rotation=45)

plt.xlabel('Período')
plt.ylabel('EAD')
plt.title('Análisis de Escenarios - Considerando Cambio de Tendencia Reciente')
plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

# ==================== 5. TABLA COMPARATIVA DE ESCENARIOS ====================
print("\n5. TABLA COMPARATIVA DE ESCENARIOS")
print("="*80)

scenarios_df = pd.DataFrame(predictions_dict)
scenarios_df['Mes'] = future_months
scenarios_df = scenarios_df[['Mes'] + list(predictions_dict.keys())]

print("\nPredicciones por Escenario:")
print(scenarios_df.round(0))

# Estadísticas resumen
print("\n\nRESUMEN ESTADÍSTICO:")
print("-"*60)
print(f"Último valor histórico (may25): {df['EAD'].iloc[-1]:.0f}")
print(f"\nPredicciones para dic25:")
for method, preds in predictions_dict.items():
    growth = ((preds[-1] / df['EAD'].iloc[-1]) - 1) * 100
    print(f"  {method:<20}: {preds[-1]:.0f} (crecimiento: {growth:+.1f}%)")

# ==================== 6. ANÁLISIS DE SENSIBILIDAD ====================
print("\n\n6. ANÁLISIS DE SENSIBILIDAD")
print("="*80)

# Analizar correlación con variables explicativas en período reciente
recent_period = df.tail(6)
corr_recent_pto = recent_period['EAD'].corr(recent_period['PTO_INV'])
corr_recent_fact = recent_period['EAD'].corr(recent_period['FACTURACION'])

print(f"Correlaciones en período reciente (últimos 6 meses):")
print(f"  EAD vs PTO_INV: {corr_recent_pto:.3f}")
print(f"  EAD vs FACTURACIÓN: {corr_recent_fact:.3f}")

# ==================== 7. RECOMENDACIONES FINALES ====================
print("\n\n7. RECOMENDACIONES BASADAS EN EL ANÁLISIS")
print("="*80)

print("""
HALLAZGOS CLAVE:
1. Se observa una clara desaceleración en el crecimiento de EAD desde dic24
2. Los últimos 3 meses muestran una tendencia casi plana
3. Los modelos tradicionales pueden sobrestimar al no capturar este cambio

ESCENARIOS RECOMENDADOS:
1. CONSERVADOR: Asumir estabilización con crecimiento mínimo (0.1% mensual)
2. MODERADO: Usar tendencia de últimos 3-4 meses con factor de desaceleración
3. OPTIMISTA: Mantener tendencia histórica pero con ajuste por ciclo

JUSTIFICACIÓN METODOLÓGICA:
- El cambio de régimen observado justifica no usar todo el histórico
- La metodología de "ventana móvil" captura mejor dinámicas recientes
- El enfoque por escenarios permite comunicar incertidumbre

MONITOREO:
- Vigilar especialmente jun25 y jul25 para confirmar tendencia
- Si EAD sigue lateral/bajando, ajustar hacia escenario conservador
- Revisar correlación con drivers (PTO_INV, FACTURACIÓN)
""")

# Crear predicción final recomendada (promedio ponderado de escenarios)
weights = {'Conservador': 0.4, 'Cambio_Regimen': 0.3, 'Tendencia_3M': 0.3}
final_prediction = np.zeros(7)

for method, weight in weights.items():
    if method in predictions_dict:
        final_prediction += np.array(predictions_dict[method]) * weight

print("\nPREDICCIÓN RECOMENDADA (Ponderación de escenarios):")
for i, month in enumerate(future_months):
    print(f"  {month}: {final_prediction[i]:.0f}")

print(f"\nCrecimiento total esperado (may25-dic25): {((final_prediction[-1]/df['EAD'].iloc[-1])-1)*100:.1f}%")
print(f"Crecimiento mensual promedio: {(((final_prediction[-1]/df['EAD'].iloc[-1])**(1/7))-1)*100:.2f}%")
